File: server/internal/protocol/binary_header.go
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
	"hash/crc32"

	"github.com/google/uuid"
)

const (
	// HeaderSize is the fixed size of the binary frame header in bytes.
	HeaderSize = 64

	// MagicNumber is the protocol magic "UD" (0x5544).
	MagicNumber uint16 = 0x5544

	// CurrentVersion is the protocol version (0x01).
	CurrentVersion uint8 = 0x01

	// MaxPayloadLength is the maximum allowed payload size for a chunk (4MB).
	MaxPayloadLength uint32 = 4 * 1024 * 1024 // 4,194,304 bytes
)

// ChunkType represents the frame type.
type ChunkType uint8

const (
	ChunkTypeData  ChunkType = 0x01 // Data payload frame
	ChunkTypeAck   ChunkType = 0x02 // Acknowledgement frame
	ChunkTypeNack  ChunkType = 0x03 // Negative acknowledgement frame
	ChunkTypeProbe ChunkType = 0x04 // Keepalive / RTT probe frame
)

// FrameFlags represents bit flags in the header.
type FrameFlags uint32

const (
	FlagEncrypted  FrameFlags = 1 << 0 // Payload is encrypted with E2EE
	FlagCompressed FrameFlags = 1 << 1 // Payload is compressed
	FlagLastChunk  FrameFlags = 1 << 2 // Final chunk of the current item
)

var (
	ErrBufferTooSmall     = errors.New("buffer is smaller than the 64-byte header size")
	ErrInvalidMagic       = errors.New("invalid protocol magic number")
	ErrUnsupportedVersion = errors.New("unsupported protocol version")
	ErrPayloadTooLarge    = errors.New("payload length exceeds maximum 4MB limit")
	ErrChecksumMismatch   = errors.New("CRC32 checksum mismatch")
)

// BinaryHeader represents the 64-byte fixed frame header.
type BinaryHeader struct {
	Magic       uint16     // Offset 0..1
	Version     uint8      // Offset 2
	ChunkType   ChunkType  // Offset 3
	SessionID   [16]byte   // Offset 4..19
	ItemIndex   uint32     // Offset 20..23
	ChunkIndex  uint32     // Offset 24..27
	TotalChunks uint32     // Offset 28..31
	PayloadLen  uint32     // Offset 32..35
	Checksum    uint32     // Offset 36..39 (CRC32-IEEE)
	Nonce       [12]byte   // Offset 40..51 (AES-256-GCM 96-bit Nonce)
	Flags       FrameFlags // Offset 52..55
	Reserved    [8]byte    // Offset 56..63
}

// NewDataHeader creates a standard data chunk header.
func NewDataHeader(sessionID uuid.UUID, itemIndex, chunkIndex, totalChunks uint32, payload []byte) BinaryHeader {
	var sessBytes [16]byte
	copy(sessBytes[:], sessionID[:])

	checksum := crc32.ChecksumIEEE(payload)
	var flags FrameFlags
	if chunkIndex+1 == totalChunks {
		flags |= FlagLastChunk
	}

	return BinaryHeader{
		Magic:       MagicNumber,
		Version:     CurrentVersion,
		ChunkType:   ChunkTypeData,
		SessionID:   sessBytes,
		ItemIndex:   itemIndex,
		ChunkIndex:  chunkIndex,
		TotalChunks: totalChunks,
		PayloadLen:  uint32(len(payload)),
		Checksum:    checksum,
		Flags:       flags,
	}
}

// NewAckHeader creates an ACK response header.
func NewAckHeader(sessionID [16]byte, itemIndex, chunkIndex uint32) BinaryHeader {
	return BinaryHeader{
		Magic:      MagicNumber,
		Version:    CurrentVersion,
		ChunkType:  ChunkTypeAck,
		SessionID:  sessionID,
		ItemIndex:  itemIndex,
		ChunkIndex: chunkIndex,
		PayloadLen: 0,
	}
}

// NewNackHeader creates a NACK response header requesting immediate retransmission.
func NewNackHeader(sessionID [16]byte, itemIndex, chunkIndex uint32) BinaryHeader {
	return BinaryHeader{
		Magic:      MagicNumber,
		Version:    CurrentVersion,
		ChunkType:  ChunkTypeNack,
		SessionID:  sessionID,
		ItemIndex:  itemIndex,
		ChunkIndex: chunkIndex,
		PayloadLen: 0,
	}
}

// Encode serializes the header into a 64-byte buffer (Big-Endian).
func (h *BinaryHeader) Encode(dst []byte) error {
	if len(dst) < HeaderSize {
		return ErrBufferTooSmall
	}

	binary.BigEndian.PutUint16(dst[0:2], h.Magic)
	dst[2] = h.Version
	dst[3] = byte(h.ChunkType)
	copy(dst[4:20], h.SessionID[:])
	binary.BigEndian.PutUint32(dst[20:24], h.ItemIndex)
	binary.BigEndian.PutUint32(dst[24:28], h.ChunkIndex)
	binary.BigEndian.PutUint32(dst[28:32], h.TotalChunks)
	binary.BigEndian.PutUint32(dst[32:36], h.PayloadLen)
	binary.BigEndian.PutUint32(dst[36:40], h.Checksum)
	copy(dst[40:52], h.Nonce[:])
	binary.BigEndian.PutUint32(dst[52:56], uint32(h.Flags))
	copy(dst[56:64], h.Reserved[:])

	return nil
}

// Decode deserializes the header from a 64-byte buffer (Big-Endian) with validation.
func DecodeBinaryHeader(src []byte) (BinaryHeader, error) {
	if len(src) < HeaderSize {
		return BinaryHeader{}, ErrBufferTooSmall
	}

	magic := binary.BigEndian.Uint16(src[0:2])
	if magic != MagicNumber {
		return BinaryHeader{}, fmt.Errorf("%w: got 0x%04x expected 0x%04x", ErrInvalidMagic, magic, MagicNumber)
	}

	version := src[2]
	if version != CurrentVersion {
		return BinaryHeader{}, fmt.Errorf("%w: got %d expected %d", ErrUnsupportedVersion, version, CurrentVersion)
	}

	var h BinaryHeader
	h.Magic = magic
	h.Version = version
	h.ChunkType = ChunkType(src[3])
	copy(h.SessionID[:], src[4:20])
	h.ItemIndex = binary.BigEndian.Uint32(src[20:24])
	h.ChunkIndex = binary.BigEndian.Uint32(src[24:28])
	h.TotalChunks = binary.BigEndian.Uint32(src[28:32])
	h.PayloadLen = binary.BigEndian.Uint32(src[32:36])
	h.Checksum = binary.BigEndian.Uint32(src[36:40])
	copy(h.Nonce[:], src[40:52])
	h.Flags = FrameFlags(binary.BigEndian.Uint32(src[52:56]))
	copy(h.Reserved[:], src[56:64])

	if h.PayloadLen > MaxPayloadLength {
		return BinaryHeader{}, fmt.Errorf("%w: %d bytes", ErrPayloadTooLarge, h.PayloadLen)
	}

	return h, nil
}

// VerifyPayloadCRC32 checks if the payload matches the header's CRC32 checksum.
func (h *BinaryHeader) VerifyPayloadCRC32(payload []byte) error {
	if uint32(len(payload)) != h.PayloadLen {
		return fmt.Errorf("payload length mismatch: header declares %d but received %d", h.PayloadLen, len(payload))
	}
	actual := crc32.ChecksumIEEE(payload)
	if actual != h.Checksum {
		return fmt.Errorf("%w: header=0x%08x actual=0x%08x", ErrChecksumMismatch, h.Checksum, actual)
	}
	return nil
}

// SessionUUID returns the SessionID formatted as a standard google/uuid.UUID.
func (h *BinaryHeader) SessionUUID() uuid.UUID {
	return uuid.UUID(h.SessionID)
}

File: client/src-tauri/src/protocol/binary_header.rs
use std::io::{self, Error, ErrorKind};
use uuid::Uuid;

pub const HEADER_SIZE: usize = 64;
pub const MAGIC_NUMBER: u16 = 0x5544; // "UD"
pub const CURRENT_VERSION: u8 = 0x01;
pub const MAX_PAYLOAD_LENGTH: u32 = 4 * 1024 * 1024; // 4MB

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum ChunkType {
    Data = 0x01,
    Ack = 0x02,
    Nack = 0x03,
    Probe = 0x04,
}

impl TryFrom<u8> for ChunkType {
    type Error = io::Error;
    fn try_from(v: u8) -> Result<Self, Self::Error> {
        match v {
            0x01 => Ok(ChunkType::Data),
            0x02 => Ok(ChunkType::Ack),
            0x03 => Ok(ChunkType::Nack),
            0x04 => Ok(ChunkType::Probe),
            _ => Err(Error::new(ErrorKind::InvalidData, format!("Unknown ChunkType: 0x{:02x}", v))),
        }
    }
}

pub const FLAG_ENCRYPTED: u32 = 1 << 0;
pub const FLAG_COMPRESSED: u32 = 1 << 1;
pub const FLAG_LAST_CHUNK: u32 = 1 << 2;

pub fn compute_crc32(data: &[u8]) -> u32 {
    let mut hasher = crc32fast::Hasher::new();
    hasher.update(data);
    hasher.finalize()
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BinaryHeader {
    pub magic: u16,
    pub version: u8,
    pub chunk_type: ChunkType,
    pub session_id: [u8; 16],
    pub item_index: u32,
    pub chunk_index: u32,
    pub total_chunks: u32,
    pub payload_len: u32,
    pub checksum: u32,
    pub nonce: [u8; 12],
    pub flags: u32,
    pub reserved: [u8; 8],
}

impl BinaryHeader {
    pub fn new_data(session_id: Uuid, item_index: u32, chunk_index: u32, total_chunks: u32, payload: &[u8]) -> Self {
        let checksum = compute_crc32(payload);
        let mut flags = 0u32;
        if chunk_index + 1 == total_chunks {
            flags |= FLAG_LAST_CHUNK;
        }

        Self {
            magic: MAGIC_NUMBER,
            version: CURRENT_VERSION,
            chunk_type: ChunkType::Data,
            session_id: *session_id.as_bytes(),
            item_index,
            chunk_index,
            total_chunks,
            payload_len: payload.len() as u32,
            checksum,
            nonce: [0u8; 12],
            flags,
            reserved: [0u8; 8],
        }
    }

    pub fn new_ack(session_id: [u8; 16], item_index: u32, chunk_index: u32) -> Self {
        Self {
            magic: MAGIC_NUMBER,
            version: CURRENT_VERSION,
            chunk_type: ChunkType::Ack,
            session_id,
            item_index,
            chunk_index,
            total_chunks: 0,
            payload_len: 0,
            checksum: 0,
            nonce: [0u8; 12],
            flags: 0,
            reserved: [0u8; 8],
        }
    }

    pub fn new_nack(session_id: [u8; 16], item_index: u32, chunk_index: u32) -> Self {
        Self {
            magic: MAGIC_NUMBER,
            version: CURRENT_VERSION,
            chunk_type: ChunkType::Nack,
            session_id,
            item_index,
            chunk_index,
            total_chunks: 0,
            payload_len: 0,
            checksum: 0,
            nonce: [0u8; 12],
            flags: 0,
            reserved: [0u8; 8],
        }
    }

    pub fn encode(&self, dst: &mut [u8]) -> io::Result<()> {
        if dst.len() < HEADER_SIZE {
            return Err(Error::new(ErrorKind::UnexpectedEof, "buffer too small for 64-byte header"));
        }

        dst[0..2].copy_from_slice(&self.magic.to_be_bytes());
        dst[2] = self.version;
        dst[3] = self.chunk_type as u8;
        dst[4..20].copy_from_slice(&self.session_id);
        dst[20..24].copy_from_slice(&self.item_index.to_be_bytes());
        dst[24..28].copy_from_slice(&self.chunk_index.to_be_bytes());
        dst[28..32].copy_from_slice(&self.total_chunks.to_be_bytes());
        dst[32..36].copy_from_slice(&self.payload_len.to_be_bytes());
        dst[36..40].copy_from_slice(&self.checksum.to_be_bytes());
        dst[40..52].copy_from_slice(&self.nonce);
        dst[52..56].copy_from_slice(&self.flags.to_be_bytes());
        dst[56..64].copy_from_slice(&self.reserved);

        Ok(())
    }

    pub fn decode(src: &[u8]) -> io::Result<Self> {
        if src.len() < HEADER_SIZE {
            return Err(Error::new(ErrorKind::UnexpectedEof, "buffer too small for 64-byte header"));
        }

        let magic = u16::from_be_bytes([src[0], src[1]]);
        if magic != MAGIC_NUMBER {
            return Err(Error::new(ErrorKind::InvalidData, format!("Invalid magic: 0x{:04x}", magic)));
        }

        let version = src[2];
        if version != CURRENT_VERSION {
            return Err(Error::new(ErrorKind::InvalidData, format!("Unsupported version: {}", version)));
        }

        let chunk_type = ChunkType::try_from(src[3])?;

        let mut session_id = [0u8; 16];
        session_id.copy_from_slice(&src[4..20]);

        let item_index = u32::from_be_bytes([src[20], src[21], src[22], src[23]]);
        let chunk_index = u32::from_be_bytes([src[24], src[25], src[26], src[27]]);
        let total_chunks = u32::from_be_bytes([src[28], src[29], src[30], src[31]]);
        let payload_len = u32::from_be_bytes([src[32], src[33], src[34], src[35]]);
        let checksum = u32::from_be_bytes([src[36], src[37], src[38], src[39]]);

        let mut nonce = [0u8; 12];
        nonce.copy_from_slice(&src[40..52]);

        let flags = u32::from_be_bytes([src[52], src[53], src[54], src[55]]);

        let mut reserved = [0u8; 8];
        reserved.copy_from_slice(&src[56..64]);

        if payload_len > MAX_PAYLOAD_LENGTH {
            return Err(Error::new(ErrorKind::InvalidData, format!("Payload too large: {} bytes", payload_len)));
        }

        Ok(Self {
            magic,
            version,
            chunk_type,
            session_id,
            item_index,
            chunk_index,
            total_chunks,
            payload_len,
            checksum,
            nonce,
            flags,
            reserved,
        })
    }

    pub fn verify_crc32(&self, payload: &[u8]) -> io::Result<()> {
        if payload.len() as u32 != self.payload_len {
            return Err(Error::new(ErrorKind::InvalidData, "Payload length mismatch"));
        }
        let actual = compute_crc32(payload);
        if actual != self.checksum {
            return Err(Error::new(ErrorKind::InvalidData, format!("CRC32 mismatch: declared=0x{:08x}, actual=0x{:08x}", self.checksum, actual)));
        }
        Ok(())
    }
}
